### ISSUE TITLE
[Clang] CodeGen issues with C++20 bitfield initializers in anonymous structs/unions

### ISSUE THREAD

**Issue by a2flo** [llvm/llvm-project#91257]
Hi,

I noticed some issues when using C++20 bitfield member initialization inside nested anonymous structs/unions. I'm not entirely sure how much of this is undefined (or intended) behavior, but it still seems to me that clangs behavior isn't right here.

Given the following code, this returns an uninitialized value from main() (compiles down to just a `ret` after optimization) and when using clang with assertions enabled, this triggers the `(Field->isUnnamedBitfield() || Field->isAnonymousStructOrUnion()) && "Only unnamed bitfields or ananymous class allowed"` assert in [CGExprAgg.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGExprAgg.cpp#L1737). Also, ananymous should be anonymous.
```cpp
struct nested_anon_bitfield_init_t {
    union {
        struct {
            int inner : 32 { 42 };
        };
        int outer;
    };
};

int main() {
    return nested_anon_bitfield_init_t{}.inner;
}
```
[on compiler-explorer](https://compiler-explorer.com/z/WoqvMdGK9)
Tested on current trunk and 18.1.
Note that GCC returns 0 here.

__

Similarly, when using less nesting, this no longer triggers the assertion, but still returns an uninitialized value:
```cpp
union less_nested_anon_bitfield_init_t {
    struct {
        int inner : 32 { 42 };
    };
};

int main() {
    return less_nested_anon_bitfield_init_t{}.inner;
}
```
[on compiler-explorer](https://compiler-explorer.com/z/e188a3dPP)
Again, GCC returns 0 here.

Intuitively, both should return 42 of course.

**Comment by efriedma-quic** [comment]
For comparison, MSVC just outright rejects this construct.

**Comment by shafik** [comment]
Note: edg(in gcc mode)/gcc are consistent here: https://compiler-explorer.com/z/dGG1bdE4e in returning zero, although I am not sure I get why.

CC @AaronBallman

**Comment by cor3ntin** [comment]
https://compiler-explorer.com/z/qrMv55frG

**Comment by a2flo** [comment]
The zero value happens due to the brace init of the struct, which just means that bitfield initializers in anonymous structs are not considered in GCC/EDG. Question is, has this been considered at all yet, since this is a GNU extension?

Different test with a standard C++ anonymous union, where clang/GCC/EDG/MSVC agree on 42:
```
struct std_anon_union_bitfield_init_t {
    union {
        int inner : 32 { 42 };
    };
};

int main() {
    return std_anon_union_bitfield_init_t{}.inner;
}
```
[on compiler-explorer](https://compiler-explorer.com/z/eo37qfMPj)

**Comment by efriedma-quic** [comment]
I'd suggest just following MSVC's lead here, rather than trying to invent our own rules for this case.

**Comment by AaronBallman** [comment]
> I'd suggest just following MSVC's lead here, rather than trying to invent our own rules for this case.

AIUI, we allow anonymous structures in C++ as an extension of the C feature, and in C, the members of the anonymous structure are members of the outer (non-anonymous) tag. From that perspective, I think it makes sense to expect `inner` to be initialized to 42. Especially because we allow it for a non-bit-field with the expected behavior: https://godbolt.org/z/98h9s7s5s

**Comment by efriedma-quic** [comment]
This was simpler than I thought; posted #91692.  But the general case gets very weird once you mix anonymous structs with anonymous unions...
